File: fs-local/src/main/java/l/files/fs/local/Passwd.java
package l.files.fs.local;

import com.google.auto.value.AutoValue;

/**
 * @see <a href="http://pubs.opengroup.org/onlinepubs/7908799/xsh/pwd.h.html">pwd.h</a>
 */
@AutoValue
abstract class Passwd extends Native {

    Passwd() {
    }

    public abstract String getName();

    public abstract int getUid();

    public abstract int getGid();

    public abstract String getDir();

    public abstract String getShell();

    public static Passwd create(String name, int uid, int gid, String dir, String shell) {
        return new AutoValue_Passwd(name, uid, gid, dir, shell);
    }

    static {
        init();
    }

    private static native void init();

    /**
     * @see <a href="http://pubs.opengroup.org/onlinepubs/7908799/xsh/getpwuid.html">getpwuid()</a>
     */
    public static native Passwd getpwuid(int uid) throws ErrnoException;

}

File: files-app/src/main/kotlin/fs/local/Passwd.kt
package l.files.fs.local

import kotlin.platform.platformStatic

/**
 * http://pubs.opengroup.org/onlinepubs/7908799/xsh/pwd.h.html
 */
private data class Passwd(val name: String,
                  val uid: Int,
                  val gid: Int,
                  val dir: String,
                  val shell: String) : Native() {

    public class object {

        {
            init()
        }

        platformStatic native private fun init()

        /**
         * http://pubs.opengroup.org/onlinepubs/7908799/xsh/getpwuid.html
         */
        throws(javaClass<ErrnoException>())
        platformStatic native fun getpwuid(uid: Int): Passwd

    }

}
